### DISCUSSION TITLE
@xstate/test testing philosophy questions

### DISCUSSION THREAD

**Discussion by ilovett** [statelyai/xstate#1311] [Q&A]
It looks like on the docs almost all the examples the state machines are supplied with some UI because the meta uses global variables like puppeteer / jest / RTL?   I see things like `await page.click('input');` but what does the DOM or supplied UI look like?

I'm having a hard time grokking how the meta `exec: async await page.click('input')` works or translates to non UI testing.

Is it possible to write test coverage on the machine itself and just `send` events to itself?  `cases` looks like it *might* be what I'm looking for -- but are there any example implementations?  And would this be done without anything in `exec` ?

I feel like I am missing the point of model based testing.

**Comment by karfau** [comment]
I also found this a bit confusing, and am still not sure if doing it this way is actually testing anything, or just validating that "xstate works as expected" but creating the test plan for a machine that is created with a function according to it's parameters, provides a nice overview of all the "tested" paths.

For that I just provided the events either with an empty object or, if needed, with cases, e.g.:

```
.withEvents({
      CHANGE: {
        cases: [{value: 'modified'}, {value: 'start'}]
      },
      VALID: {},
      INVALID: {},
      SUBMIT: {}
    });
```

Does that answer your question in any way?

**Reply by ilovett** [reply]
Thanks but not really, I found there are too many gaps to fill in the documentation so I gave up.  I don't really see how the `cases` will be used: ` the sample event objects for this event type that can be sent by the testing model.` -- what does this mean?  How does the testing model work?  Can we get some kind of explanation of sequence diagram of how it works?

Ultimately I gave up on `@xstate/test` and wrote my own specific test cases using `withConfig` for mocking and `withContext`, and passing my starting state at `.start({ ... })`... seems way more straight forward.

**Reply by davidkpiano** [reply]
> How does the testing model work? Can we get some kind of explanation of sequence diagram of how it works?

Thanks for the feedback; I can do a better job of explaining this in the docs.

The way `@xstate/test` works is essentially by doing a depth-first search of the statechart, where, given state `first`, the next state would be `machine.transition(first, someEvent)`. Of course, for any state, a different event can occur, so it will try `machine.transition(first, anotherEvent)` on another pass.

And sometimes, events might lead to different states given different payloads. So instead of just sending `{ type: 'LOGIN' }`, you might want to send `{ type: 'LOGIN', username: 'validUser' }` and `{ type: 'LOGIN', username: 'invalidUser' }` etc. to try the different paths. Those events are _cases_.

**Comment by rjdestigter** [comment]
The UI in the examples is just the _system under test_ (SUT). But it doesn't have to be. It could be a web-server, or a calculator, or a state machine, anything really.

State machines in combination with _@xstate/test_'s _createModel_ API are just tools here to define your system in terms of a state machine. You define the states your system can be in and write tests for those states. Additionally you describe events that can happen in each state and what code should be executed to transition your system to the next testable state.

You're observing your system from the outside and asking yourself: "Okay, here's my system/ui/program/server/anything, this is it's initial state.". Next question is: "What events can occur?" Can you click a button. Or can you execute a command in the shell that will start the server. Or would you execute some code. That's how you build your model. XState will build a machine for you and xstate/test will figure out all the pathways and execute the tests and events in the right and possible orders.

**Reply by karfau** [reply]
> or a state machine

Since this was what I was trying to do, what I was not able to grasp:
When testing a state machine, wouldn't I just create the model from the machine under test and define all the events possible in the machine? 
I did that (see my comment above, without any `exec` or `meta.test`) and now I have some auto-generated "green" tests but What would such a test actually "prove"?

**Reply by Silverwolf90** [reply]
You are right that it proves nothing, all it did was generate ways to traverse your state machine -- not whether these traversals were "correct" based on your expectations. This is why you need a different state machine to test your state machine. In other words, a new state machine that models how you might expect one to "use" your initial state machine.

**Reply by karfau** [reply]
Wouldn't such a state machine have the same (or very similar) states (and the same events) as the machine under test?
I'm still having a hard time to have any idea howto come up with such a machine and I would be happy about any examples. (Just to make it clear: I have a rough idea how to do that when a UI is present.)

**Reply by Andarist** [reply]
This is something that I have a hard time to grasp as well - I understand the concept alone (of having 2 separate models - the implementation one and the test one) but it's hard for me to apply this in the real world scenario when my implementation is already using a machine 🤔 

This comes quite often as a question - would be great to have a good example of this in the docs, cc @davidkpiano

**Reply by davidkpiano** [reply]
> Wouldn't such a state machine have the same (or very similar) states (and the same events) as the machine under test?

Not exactly!

Think about ordering food from a restaurant. To you, the customer, the process goes like this:

1. A waiter asks you what you want to eat
2. You tell them
3. The waiter takes your order... and you wait.
4. Eventually, you get your food!

But what's really happening?

1. A waiter is assigned to your section, but must finish the queue of tables that they're working on already.
2. Once done, the waiter comes over and asks you what you want to eat.
3. The waiter receives your order, and writes it down
4. The waiter dispatches the order to the kitchen
5. The kitchen must go through a queue of previous orders before getting to your order
6. The order is finally picked up and prepared by the kitchen
7. Once all the food is cooked (this might be several steps!), the food is plated
8. The plated food is waiting to be picked up by the waiter
9. Once the waiter is able to grab your food, they do and bring it to your table
10. The food is delivered to your table

You can consider these two workflows of vastly differing complexity! The first "state machine" is much simpler than the second "state machine".

So which state machine should you test? The actual, complex state machine? Or the state machine from the user's perspective?

In behavioral testing (also known as black-box testing), you should be testing the **simpler state machine** - the one that represents _what is externally happening_. Often, this is different than the actual state machine, which represents implementation details and details that aren't visible to the user.

From a software perspective, this example can be made more real if you consider that when you submit a form, you don't know what's happening. You just know that there's a loading spinner, and eventually, you'll be taken to the next screen (which hopefully shows a successful form submission).

But in reality, there can be a lot happening behind that loading spinner. It doesn't matter to the user, though. And it shouldn't be tested in E2E tests.

Hope this makes sense!

**Reply by Andarist** [reply]
If the "inner" logic has different branches of logic - how to test those if we only focus on the simpler machine? Or are you saying that testing that is just not a job for `@xstate/test`?

**Reply by rjdestigter** [reply]
> Or are you saying that testing that is just not a job for @xstate/test?

That's what I would say. I think this is where the confusion lies. Developers make the assumption that _@xstate/test_ was made for testing the state machines they develop for their application. (I did). The light bulb turned on when I realised _@xstate/test_ was made for creating and automating model based tests regardless of what the _system under test_ is.

**Reply by rjdestigter** [reply]
You can still use it to test state machines. But you should observe the machine's behaviour from the outside as David's example scenario describes.

**Reply by Andarist** [reply]
I understand both things - have been definitely confused at first by the distinction between the SUT (that can be machine-based but doesn't have to be) and the machine used with `@xstate/test` but that's no longer the case. However, I still fail to understand how to implement this correctly in practice.

I'm not sure if I've been understood correctly by you (I could have been, but I'm not 100% sure) so let me reiterate my question. I'm asking if `@xstate/test` is a good tool to test the inner complexities of the system - this actually still doesn't have to be implemented using machines but it might be. While often the observable effect is what we want to actually test it can be achieved using different logic paths (often massively different) and I'm wondering if those can be tested using this approach. If they can then I would imagine that the testing model would become closer to the SUT model with each requirement covered by the testing model and then I get lost because it leads to both being implemented in a very similar way (assuming SUT is implemented using machines).

**Reply by rjdestigter** [reply]
I think I understand what you are wondering about. In my project at work I did initially write some model based tests for _inner complexities_, primarily for state machines used in the application. I eventually dropped that though because it didn't feel like it was adding anything. With TypeScript set to `strict: true` I don't even write unit tests anymore.

What I do try to do with my integration tests is throw all possible input variants at the system. Since the system doesn't really do anything unless a network request is resolving or a user is typing or clicking, I think writing integration tests that just test the observed expected behaviour implicitly covers what is happening on the inside. This is also part of the goals I have for marionettist. To be able to easily define the outcome of (asynchronous) internal events such as network requests, timeout/interval outcomes, etc. If those are covered plus the user driven events then you get a more complete test of the entire system.

**Reply by davidkpiano** [reply]
Nothing is stopping you from using `@xstate/test` to test your actual "inner" complex state machines. Not sure where the confusion is here - you can use the library however you like.

**Reply by rjdestigter** [reply]
> Not sure where the confusion is here - you can use the library however you like.

That's exactly the confusion for some. New comers often assume _@xstate/test_ is specifically for testing state machines.

**Reply by davidkpiano** [reply]
Hmm, I should write a testing guide in the docs. Testing machines is straightforward, whether you do it in a "stateless" manner:

```js
it('should transition to the right state', () => {
  // ...
  expect(machine.transition(someState, someEvent).value).toEqual(expected);
});
```

Or through a service ("async"):

```js
it('should end up in the correct state', done => {
  const events = [/* ... */];
  const expectedState = // ... e.g., "success"

  const service = interpret(someMachine)
    .onTransition(state => {
      if (state.matches(expectedState)) { done(); }
    })
    .start();

  events.forEach(event => service.send(event));
});
```

And it doesn't require a library. `@xstate/test` is for model-based testing.

**Reply by karfau** [reply]
I really think #1601 is a really good summary of this thread (and I guess it will soon be available under <https://xstate.js.org/docs/guides/testing.html>).
But the previous comment is just spot on.

Thx a lot for all the clarifying comments.

**Reply by karfau** [reply]
I found one more place where the description of the package was ambiguous and filed #1613

**Comment by davidkpiano** [comment]
Documentation updated with testing: https://xstate.js.org/docs/guides/testing.html

**Comment by karfau** [comment]
I have another question (that fit's the very broad title):

When writing model based tests, is it common/uncommon to make use of the context?

On the one hand I can imagine that it would be possible to model conditional behaviour of the SUT by modelling everything using different event types and states, but there would of course be some redundancy.

On the other hand, when using context to make assertions based on it in `meta.test` it's not quite clear to me, how to manage the context in between test with the different states and event involved, should there be many `onEntry` actions to `assign` the related values? (Since there is only a single machine instance used for all describes in a model.)

Any examples are very welcome.

**Reply by davidkpiano** [reply]
> When writing model based tests, is it common/uncommon to make use of the context?

It's common. The context _is_ a part of the state (the state is a combination of the context and the finite state value when mapping out the overall app states).

**Reply by karfau** [reply]
Thx, I would love to better understand how to use the context in model based testing, are you aware of some open source examples I could look at?

**Reply by davidkpiano** [reply]
Take a look at the "Die Hard Puzzle" test: https://github.com/davidkpiano/xstate/blob/99c7311edb1032bff81358d84fd7f41c6cb7f657/packages/xstate-test/test/index.test.ts#L6-L170

**Reply by karfau** [reply]
From that example I found that the second argument passed to `state.meta.test` is the `state`, which is a very helpful information indeed. I tried to verify this by searching in the docs, but couldn't find any references. Are you aware of any?
